### PR TITLE
dockershim: set security option separators based on the docker version

### PR DESCRIPTION
**Pull request by yujuhong**
Also add a version cache to avoid hitting the docker daemon frequently.

This is part of #38164

### CODE REVIEW

**Comment by yujuhong** [comment]
This PR is based on #40903. Only the latest commit is new.

**Comment by yujuhong** [comment]
@feiskyer @Random-Liu could you help review the PR? Thanks!

**Comment by feiskyer** [comment]
Reviewing

**Comment by feiskyer** [comment]
LGTM

**Comment by yujuhong** [comment]
Removing the lgtm label since this PR depends on #40903, which is still under review.

**Comment by yujuhong** [comment]
@Random-Liu I'll rebase once #40903 is merged, but I believe it should be trivial, so feel free to review the second commit. Thanks!

**Comment by Random-Liu** [comment]
@yujuhong Will review it today. :)

File: pkg/kubelet/dockershim/docker_service.go [hunk 1]
```diff
@@ -100,7 +105,12 @@ type NetworkPluginSettings struct {
 	LegacyRuntimeHost network.LegacyHost
 }
 
-var internalLabelKeys []string = []string{containerTypeLabelKey, containerLogPathLabelKey, sandboxIDLabelKey}
+var (
+	internalLabelKeys []string = []string{containerTypeLabelKey, containerLogPathLabelKey, sandboxIDLabelKey}
+
+	// The expiration time of version cache.
+	versionCacheTTL = 60 * time.Second
```
Comment (Random-Liu) [inline pkg/kubelet/dockershim/docker_service.go]: Why not const?

File: pkg/kubelet/dockershim/docker_service_test.go [hunk 2]
```diff
@@ -90,3 +104,25 @@ func TestStatus(t *testing.T) {
...
+	v, err := ds.getDockerVersion()
+	require.NoError(t, err)
+	assert.Equal(t, expectedVersion, v)
+
+	expectedAPIVersion := &semver.Version{Major: 1, Minor: 23, Patch: 0}
+	apiVersion, err := ds.getDockerAPIVersion()
+	require.NoError(t, err)
+	assert.Equal(t, expectedAPIVersion, apiVersion)
+}
+func TestAPIVersionWithCache(t *testing.T) {
```
Comment (Random-Liu) [inline pkg/kubelet/dockershim/docker_service_test.go]: Add a new line.

File: pkg/kubelet/dockershim/docker_service.go [hunk 3]
```diff
@@ -298,3 +326,33 @@ func (ds *dockerService) GenerateExpectedCgroupParent(cgroupParent string) (stri
...
+	if err != nil {
+		return nil, err
+	}
+	return &apiVersion, nil
+}
+
+func (ds *dockerService) getDockerVersionFromCache() (*dockertypes.Version, error) {
+	const (
+		// We only store on key in the cache.
+		dummyKey = "version"
```
Comment (Random-Liu) [inline pkg/kubelet/dockershim/docker_service.go]: Why not `const dummyKey = "version"`?

File: pkg/kubelet/dockershim/docker_service_test.go [hunk 4]
```diff
@@ -40,9 +44,19 @@ func newTestNetworkPlugin(t *testing.T) *mock_network.MockNetworkPlugin {
 
 func newTestDockerService() (*dockerService, *dockertools.FakeDockerClient, *clock.FakeClock) {
 	fakeClock := clock.NewFakeClock(time.Time{})
-	c := dockertools.NewFakeDockerClient().WithClock(fakeClock)
-	return &dockerService{client: c, os: &containertest.FakeOS{}, networkPlugin: &network.NoopNetworkPlugin{},
-		legacyCleanup: legacyCleanupFlag{done: 1}, checkpointHandler: NewTestPersistentCheckpointHandler()}, c, fakeClock
+	c := dockertools.NewFakeDockerClient().WithClock(fakeClock).WithVersion("1.11.2", "1.23")
+	return &dockerService{client: c, os: &containertest.FakeOS{}, networkPlugin: &network.NoopNetworkPlugin{}, legacyCleanup: legacyCleanupFlag{done: 1}, checkpointHandler: NewTestPersistentCheckpointHandler()}, c, fakeClock
```
Comment (Random-Liu) [inline pkg/kubelet/dockershim/docker_service_test.go]: Why remove the newline?

File: pkg/kubelet/dockershim/docker_service_test.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/kubelet/dockershim/docker_service_test.go]: What do we want to test with this function? Should we check fake docker client to see whether `version` is actually cached?

File: pkg/kubelet/dockershim/helpers_test.go [hunk 5]
```diff
@@ -236,3 +237,28 @@ func TestParsingCreationConflictError(t *testing.T) {
...
+		"older docker version": {
+			version:  &semver.Version{Major: 1, Minor: 22, Patch: 0},
+			expected: ':',
+		},
+		"changed docker version": {
+			version:  &semver.Version{Major: 1, Minor: 23, Patch: 0},
+			expected: '=',
+		},
+		"newer docker version": {
+			version:  &semver.Version{Major: 1, Minor: 23, Patch: 0},
```
Comment (Random-Liu) [inline pkg/kubelet/dockershim/helpers_test.go]: 24?

File: pkg/kubelet/dockershim/helpers_test.go [hunk 6]
```diff
@@ -236,3 +237,28 @@ func TestParsingCreationConflictError(t *testing.T) {
...
+			expected: '=',
+		},
+	} {
+		actual := getSecurityOptSeparator(test.version)
+		assert.Equal(t, test.expected, actual, c)
+	}
+
```
Comment (Random-Liu) [inline pkg/kubelet/dockershim/helpers_test.go]: Extra new line.

Comment (yujuhong) [inline pkg/kubelet/dockershim/helpers_test.go]: done

File: pkg/kubelet/dockershim/helpers_test.go (hunk 5 above)
Comment (yujuhong) [inline pkg/kubelet/dockershim/helpers_test.go]: fixed

File: pkg/kubelet/dockershim/docker_service_test.go (hunk 2 above)
Comment (yujuhong) [inline pkg/kubelet/dockershim/docker_service_test.go]: done

File: pkg/kubelet/dockershim/docker_service_test.go (hunk 4 above)
Comment (yujuhong) [inline pkg/kubelet/dockershim/docker_service_test.go]: Lost in rebase. Added it back.

File: pkg/kubelet/dockershim/docker_service.go (hunk 3 above)
Comment (yujuhong) [inline pkg/kubelet/dockershim/docker_service.go]: done.

File: pkg/kubelet/dockershim/docker_service.go (hunk 1 above)
Comment (yujuhong) [inline pkg/kubelet/dockershim/docker_service.go]: a honest mistake. moved.

**Comment by yujuhong** [comment]
Addressed all comments. Since most of them are minor changes, I didn't create a new commit.

**Comment by yujuhong** [comment]
@k8s-bot gce etcd3 e2e test this

**Pull request merged**